src/services/pcn-packetcapture/src/Packetcapture_dp_egress.c

```diff
@@ -0,0 +1,198 @@
+  if(!status){
+      return RX_DROP;
+  }
+  
+  if(*status == OFF){
+      return RX_OK;
+  }
+
+  struct packetHeaders *pkt;
+  pkt = pkt_header.lookup(&key);
+  if (pkt == NULL) {
+    return RX_DROP;
+  }
+  
+  /* Parsing L2 */
+  void *data = (void *)(long)ctx->data;
+  void *data_end = (void *)(long)ctx->data_end;
+  struct eth_hdr *ethernet = data;
+  if (data + sizeof(*ethernet) > data_end)
+    return RX_DROP;
+
+  pkt->srcMac = ethernet->src;
+  pkt->dstMac = ethernet->dst;
+  uint16_t ether_type = ethernet->proto;
+
+  pkt->vlan_present = ctx->vlan_present;
+  if (pkt->vlan_present) {
+    ether_type = ctx->vlan_proto;
+    pkt->vlan = (uint16_t)(ctx->vlan_tci & 0x0fff);
+  }
+
+  /* Parsing L3 */
+  struct iphdr *ip = NULL;
+  struct tcphdr *tcp = NULL;
+  struct udphdr *udp = NULL;
+  pkt->ip = 0;
+
+  if (ether_type == bpf_htons(ETH_P_IP)) {
+    ip = data + sizeof(*ethernet);  
+    if (data + sizeof(*ethernet) + sizeof(*ip) > data_end)
+      return RX_DROP;   
+    if((int)ip->version != 4){
+      return RX_OK;
+    }
+
+    struct filters_table *filters_tab;
+    filters_tab = filters_map.lookup(&key);
+    if (filters_tab == NULL) {
+      return RX_DROP;
+    }
+    /* Souce Ip filter */
+    if((filters_tab->network_filter_src_flag == true) && ((filters_tab->netmask_filter_src & ntohl(ip->saddr)) != filters_tab->network_filter_src)){
+      return RX_OK;
+    }
+      
+    /* Destination Ip filter */
+    if((filters_tab->network_filter_dst_flag == true) && ((filters_tab->netmask_filter_dst & ntohl(ip->daddr)) != filters_tab->network_filter_dst)){
+      return RX_OK;
+    }
+
+    /* l4proto Filter */
+    if(filters_tab->l4proto_flag == true){
+      if((filters_tab->l4proto_filter == 1) && (ip->protocol != IPPROTO_TCP)){         /* tcp only */
+        return RX_OK;
+      }else if((filters_tab->l4proto_filter == 2) && (ip->protocol != IPPROTO_UDP)){   /* udp only */
+        return RX_OK;
+      }
+    }
+
+    pkt->ip = 1;
+    pkt->srcIp = ip->saddr;
+    pkt->dstIp = ip->daddr;
+    if (ip->protocol == IPPROTO_TCP) {
+      tcp = data + sizeof(*ethernet) + sizeof(*ip);
+      if (data + sizeof(*ethernet) + sizeof(*ip) + sizeof(*tcp) > data_end)
+        return RX_DROP;
+      pkt->l4proto = IPPROTO_TCP;
+
+      /* src port filter */
+      if((filters_tab->src_port_flag == true) && (filters_tab->src_port_filter != ntohs(tcp->source))){
+        return RX_OK;
+      }
+
+      /* dst port filter */
+      if((filters_tab->dst_port_flag == true) && (filters_tab->dst_port_filter != ntohs(tcp->dest))){
+        return RX_OK;
+      }
+
+      pkt->srcPort = tcp->source;
+      pkt->dstPort = tcp->dest;
+    } else if (ip->protocol == IPPROTO_UDP) {
+      udp = data + sizeof(*ethernet) + sizeof(*ip);
+      if (data + sizeof(*ethernet) + sizeof(*ip) + sizeof(*udp) > data_end)
+        return RX_DROP;
+      pkt->l4proto = IPPROTO_UDP;
+      pkt->srcPort = udp->source;
+      pkt->dstPort = udp->dest;
+    }
+
+  }
```

Comment:
Looks you send all the traffic to the controller, shouldn't we let the non-IP pass through?

Reply:
in this case we would lose the non-IP traffic in the capture, like ARP messages. 
I thought it useful not to discard that type of traffic

Reply:
if we want to capture the non-ip packet, we should probably specify it in the filter by adding some protocol support like ARP(or at least we have some knob to collect all the non-IP protocol packet), otherwise we will get all the non-IP packet in controller, which I think bring a couples of issues: 1. performance; 2. break the protocol. Think about when you enable a interface packet capture feature if the interface has IPSEC traffic, OSPF, IGMP etc...

Reply:
Ok, then we should add a feature improvement issue.